lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js

```diff
@@ -181,6 +181,232 @@
         }
     });
 
+    platform.registerChakraLibraryFunction("DefaultMergeSort", function(array, length) {
+        const compareArray = [];
+        for (let index = 0; index < length; ++index) {
+            __chakraLibrary.arrayCreateDataPropertyOrThrow(compareArray, index, "" + array[index]);
+        }
+
+        const fast = length < 1073741825;
+        const buffer = [];
+        const stringBuffer = [];
+
+        let i = 1, j = 0, k = 0, bucketSize = 4, position = 0;
+
+        // handle pairs first
+        while (i < length) {
+            if (compareArray[i] < compareArray[i - 1]) {
+                const cache = array[i];
+                array[i] = array[i - 1];
+                array[i - 1] = cache;
+
+                const stringCache = compareArray[i];
+                compareArray[i] = compareArray[i - 1];
+                compareArray[i - 1] = stringCache;
+            }
+            i += 2;
+        }
+        const doubleLength = length * 2;
+
+        while (bucketSize < doubleLength) {
+            while (position < length) {
+                const left = position;
+                let right = position + bucketSize;
+                right < length ? right : length;
```

Comment:
what does this line do?

Reply:
That looks like a no-op; it evaluates to either `right;` or `length;` which will have no effect.  Remnant from refactoring maybe?

Reply:
Without this you walk off the end of the array unless its length is a power of 2. Perhaps I should add comments explaining the algorithm?

Example - array of length 7:
1. step 1 - the first loop does pairs:
0-1, 2-3, 4-5 (you'll need it actually runs with i = 1 -> i < length always looking at i and i-1 so no risk of walking off the array)

2. step 2 - bucket size = 4:
0-4, 5-8
BUT if you run with right = 8 you would access array[7] which doesn't exist - this condition limits it so you look at 5-7 instead which is fine>

3. step 3 - bucket size = 8:
0-8
Again you need right capped at length which is 7.

Reply:
The thing is, you have a bare ternary here:
```js
// this is the entire statement
right < length ? right : length;
```

Which is basically the same thing as writing
```js
0;
```

Did you mean to assign the result of the ternary to something?

Reply:
Ok yes it's an error It's meant to be:  ```right = right < length ? right : length;```

I'm confused that this doesn't a fail a test with this error here. I'm either covering the same case with a condition somewhere else - which I cannot currently see or there's no array in a test case with a length that isn't a power of 2/

Reply:
I'm hoping that fixing this issue will also improve performance, because indexing out of bounds on arrays is very slow, and putting `undefined` inputs into the comparator function is likely deoptimizing it.

Reply:
Fixing this seems to show around a 10-15% performance improvement.

Reply:
My tests don't show any consistent improvement - though I imagine that numerous sorts of small arrays would show an improvement whereas I have been testing with sorting just 6 large arrays.